src/lib.rs

```diff
@@ -2943,11 +2943,39 @@ impl Connection {
         {
             if let Some(max_dgram_payload) = self.dgram_max_writable_len() {
                 while let Some(len) = self.dgram_send_queue.peek_front_len() {
-                    if (len + frame::MAX_DGRAM_OVERHEAD) <= left {
+                    let hdr_off = b.off();
+                    let hdr_len = 1 + // frame type
+                        2; // length, always encode as 2-byte varint
+
+                    if (hdr_len + len) <= left {
                         // Front of the queue fits this packet, send it
                         match self.dgram_send_queue.pop() {
                             Some(data) => {
-                                let frame = frame::Frame::Datagram { data };
+                                let (mut dgram_hdr, mut dgram_payload) =
```

Comment:
I found this hard to understand first of all. The comments in the STREAM frame writing also apply here, which helped my understanding. It might be overkill to copy all of the commentary but spelling out that you start by writing the payload and then jump back to write the header would really help

Reply:
I copied some of the comments, might as well copy all of them :D

Reply:
thanks!